### PR TITLE
[ENH] Detect and install missing add-ons when when loading a workflow

### PR DESCRIPTION
**Pull request by ales-erjavec**
#### Issue

Loading a workflow containing nodes from add-ons fails if the add-ons are not installed.

#### Changes

Detect missing packages and offer to install them via the add-ons dialog.

### CODE REVIEW

File: orangecanvas/application/canvasmain.py [hunk 1]
```diff
@@ -1179,10 +1248,87 @@ def new_scheme_from(self, filename):
...
+                     "Would you like to install them now?",
+                standardButtons=QMessageBox.Ok | QMessageBox.Abort |
+                                QMessageBox.Ignore,
+                informativeText=(
+                    "After installation you will have to restart the "
+                    "application and reopen the workflow"),
+            )
+            mb.setDefaultButton(QMessageBox.Ok)
+            bignore = mb.button(QMessageBox.Ignore)
+            bignore.setText("Load Partial")
```
Comment (ajdapretnar) [inline orangecanvas/application/canvasmain.py]: I was a bit confused by "Load Partial" button. The tooltip explains it nicely, but I was thinking whether there's a more informative label for it. Perhaps there isn't.

**Comment by ajdapretnar** [comment]
This works well. Just a couple of observations:

1.) If I select Load Partial, I still get the Error Message. I don't expect it to, since I have willingly selected Load Partial workflow.

2.) Formatting of the prompt. The buttons should probably extend to the window width.
Screen Shot 2019-11-22 at 10 20 16

3.) Formatting of the installation window.
- Filter search and 'Add more...' are probably redundant? (It seems that this is a standard add-on dialogue, but could we at least disable them for this case? 🤔 )
- Version column too narrow.
Screen Shot 2019-11-22 at 10 21 06

File: orangecanvas/application/canvasmain.py [hunk 2]
```diff
@@ -1152,25 +1181,65 @@ def new_scheme_from(self, filename):
...
-                )
-        except Exception:
+            new_scheme.load_from(
+                fileobj, registry=self.widget_registry,
+                error_handler=errors.append
+            )
+        except Exception:  # pylint: disable=broad-except
+            log.exception("")
             message_critical(
                  self.tr("Could not load an Orange Workflow file"),
```
Comment (ajdapretnar) [inline orangecanvas/application/canvasmain.py]: I suggest we treat this as a sentence and finish with a full stop.

File: orangecanvas/application/canvasmain.py [hunk 3]
```diff
@@ -1179,10 +1248,87 @@ def new_scheme_from(self, filename):
...
+                icon=QMessageBox.Question,
+                windowTitle="Install Additional Packages",
+                text="This workflow requires additional packages to run. "
+                     "<br/>" + details + "<br/>"
+                     "Would you like to install them now?",
+                standardButtons=QMessageBox.Ok | QMessageBox.Abort |
+                                QMessageBox.Ignore,
+                informativeText=(
+                    "After installation you will have to restart the "
+                    "application and reopen the workflow"),
```
Comment (ajdapretnar) [inline orangecanvas/application/canvasmain.py]: Same as above. I would finish with a full stop.

**Comment by ajdapretnar** [comment]
After some discussion we propose to have a text like this:
Workflow you are trying to load contains widgets from missing add-ons.

[Ignore missing widgets]
[Abort]
[Install add-ons]

Did I get it right? @markotoplak @janezd

**Comment by ales-erjavec** [comment]
> 1.) If I select Load Partial, I still get the Error Message. I don't expect it to, since I have willingly selected Load Partial workflow.

The message reports all missing node types and links that are omitted. This can happen for either missing installed packages or e.g. incompatible installed package versions. There is (currently) no easy way to distinguish between them. Also the actual set of deficiencies in the loaded workflow should in some way be reported to the user even when choosing to load only the partial workflow.

> 3.) Formatting of the installation window.

Fixed.

**Pull request merged**